src/HDWallet.cpp

```diff
@@ -319,31 +321,36 @@ HDNode getNode(const HDWallet& wallet, TWCurve curve, const DerivationPath& deri
     auto node = getMasterNode(wallet, curve);
     for (auto& index : derivationPath.indices) {
         switch (privateKeyType) {
-            case HDWallet::PrivateKeyTypeDoubleExtended: // used by Cardano, special handling
-                hdnode_private_ckd_cardano(&node, index.derivationIndex());
-                break;
-           case HDWallet::PrivateKeyTypeDefault32:
-            default:
-                hdnode_private_ckd(&node, index.derivationIndex());
-                break;
+        case HDWallet::PrivateKeyTypeDoubleExtended:
+            hdnode_private_ckd_cardano(&node, index.derivationIndex());
+            break;
+        case HDWallet::PrivateKeyTypeDefault32:
+        default:
+            hdnode_private_ckd(&node, index.derivationIndex());
+            break;
         }
     }
     return node;
 }
 
 HDNode getMasterNode(const HDWallet& wallet, TWCurve curve) {
     const auto privateKeyType = HDWallet::getPrivateKeyType(curve);
-    auto node = HDNode();
+    HDNode node;
     switch (privateKeyType) {
-        case HDWallet::PrivateKeyTypeDoubleExtended: // used by Cardano
-            // special handling for extended, use entropy (not seed)
-            hdnode_from_entropy_cardano_icarus((const uint8_t*)"", 0, wallet.getEntropy().data(), (int)wallet.getEntropy().size(), &node);
-            break;
-
-        case HDWallet::PrivateKeyTypeDefault32:
-        default:
-            hdnode_from_seed(wallet.getSeed().data(), HDWallet::seedSize, curveName(curve), &node);
-            break;
+    case HDWallet::PrivateKeyTypeDoubleExtended: {
+        // Derives the root Cardano HDNode from a passphrase and the entropy encoded in
+        // a BIP-0039 mnemonic using the Icarus derivation (V2) scheme
+        const auto entropy = wallet.getEntropy();
+        uint8_t secret[CARDANO_SECRET_LENGTH];
+        secret_from_entropy_cardano_icarus((const uint8_t*)"", 0, entropy.data(), int(entropy.size()), secret, nullptr);
+        hdnode_from_secret_cardano(secret, &node);
+        memzero(secret, CARDANO_SECRET_LENGTH);
```

Comment:
here you use memzero function from trezor, which is a wrapper around `std::memset`. How about using `memzero` everywhere for convenience? If the type is a [pod](https://en.cppreference.com/w/cpp/named_req/PODType) so it's should be perfectly accepted:

```cpp
namespace {
    template<typename T>
    static inline constexpr void zeroInit(T& data, std::size_t len = sizeof(T)) noexcept {
         static_assert(std::is_pod_v<T>, "type should be a pod");
         memzero(&data, len);
    }
}
```

Reply:
looks good, can I do it in another pr? this pr scope gets bigger and bigger

Reply:
Sure

Reply:
tracked here https://github.com/trustwallet/wallet-core/issues/2386